lambdas/shared/t4_lambda_shared/preview.py

```diff
@@ -129,81 +129,80 @@ def extract_fcs(file_, as_html=True):
     # in previous versions (git blame) we sent a lot more schema information
     # but it's heavy on the browser and low information; just send column names
-    info['schema'] = {
-        'names': meta.schema.names
-    }
-    info['serialized_size'] = meta.serialized_size
+    info['schema'] = {'names': meta.schema.names}
+    info['serialized_size'] = meta.serialized_size  # footer serialized size
     info['shape'] = [meta.num_rows, meta.num_columns]
-    # TODO: refactor preview code to use dask/s3fs and pyarrow.dataset scanner to
-    # spare memory; part of the reason this is so inefficient: we've already read
-    # the entire parquet file into a BytesIO by the time we get here
-    if meta.num_row_groups:
-        # guess because we meta doesn't reveal how many rows in first group
-        num_rows_guess = math.ceil(meta.num_rows / meta.num_row_groups)
-        size_guess = num_rows_guess * meta.num_columns * AVG_PARQUET_CELL_BYTES
-        if skip_rows or (size_guess > get_available_memory()):
-            # minimal dataframe with all columns and one row
-            dataframe = pandas.DataFrame(columns=meta.schema.names)
-            info['warnings'] = 'Large file: skipped rows to conserve memory, only showing column names'
-        else:
-            dataframe = pf.read_row_group(0)[:MAX_PREVIEW_ROWS].to_pandas()
-    # sometimes there are neither rows nor row_groups, just columns
-    # therefore we do not call read_row_group because (with 0 row_groups)
-    # it would barf
-    else:
-        # :0 is a safety valve but there really should be no rows in this case
-        dataframe = pf.read()[:0].to_pandas()
-    if as_html:
-        body = remove_pandas_footer(dataframe._repr_html_())  # pylint: disable=protected-access
+
+    available = get_available_memory()
+    iter_batches = None
+    # 10MB heuristic; should never happen, e.g. with current default of 512MB
+    if (available < 10E6) or skip_rows:
+        logger_.warning("Insufficient memory to index parquet file: %s", info)
+        info['warnings'] = "Skipped rows; insufficient memory"
+    elif meta.num_rows and meta.num_row_groups:
+        iter_batches = pf.iter_batches(batch_size=128, row_groups=[0])
     else:
+        logger_.warning("Parquet file with no rows: %s", info)
+    if iter_batches:
         buffer = []
         size = 0
         done = False
-        for _, row in dataframe.iterrows():
-            for column in row.astype(str):
-                encoded = column.encode()
-                # +1 for \t
-                encoded_size = len(encoded) + 1
-                if (size + encoded_size) < max_bytes:
-                    buffer.append(encoded)
-                    buffer.append(b"\t")
-                    size += encoded_size
-                else:
-                    done = True
-                    break
-            buffer.append(b"\n")
-            size += 1
+        for batch in iter_batches:
+            df = batch.to_pandas()
```

Comment:
If I remember correctly, `to_pandas` is useful for the `_repr_html_()` method - but do we still need it in the non-HTML case? If you can access parquet values directly, that's probably more efficient. But I don't know much about all this.

Reply:
Good question. Let me think about it. Not sure if the resulting effort will save us memory in that ostensibly arrow has a very efficient way to convert to Pandas. Is there anything I can do to make sure batches get garbage collected better?

Reply:
Ok. Anyways, let's leave it alone - I was curious, but not worth it trying to rewrite everything at once. Garbage collection, etc. should be fine.